### PR TITLE
Evals: add quality validation flow to create/update eval skills

### PR DESCRIPTION
**Pull request by nehaoss**
**Why**
The eval create and update flows had no quality gate — generated test cases were pushed without any review of whether they were realistic, diverse, or well-formed. When a gate did exist, the fix flow was clunky (per-case Y/N/Skip for every flagged case).

**How it works**

Quality validation uses a parent + subagent split:

- The validate subagent (validate/SKILL.md) is a stateless reviewer — it reads the eval files, applies the 9 dimension rubric, and returns a structured quality report. It has no knowledge of the create/update flow.

- The parent agent (create/update skill) invokes the subagent via **runSubagent** (inbuilt from vs code), displays its full report verbatim, then decides what to do next based on the gate result:
1. Pass (≥ 4/5) → proceed to review
2. Review/Fail (≤ 3/5) → run batch fix flow → re-invoke subagent → display new report → proceed to review

The subagent is re-invoked fresh after fixes (no shared state), so the second score is an independent re-evaluation, not a continuation of the first.

image

**What changed**

**New**: validate skill

Validate subagent that scores eval test sets across 9 quality dimensions (Validity, Realism, Assertion Quality, Coverage, Diversity, Redundancy, Failure Mode Coverage, Discriminative Power, Topic Alignment)
Outputs a dimension table with per-category scores and ⚠️ callouts for flagged cases
Pass ≥ 4/5 · Review = 3/5 · Fail ≤ 2/5

**Updated**:  create skill

Step 4.3: invoke validate subagent after writing files, display full dimension report verbatim
Fix flow: apply all (or selected) fixes in batch → show Before/After summary table → re-validate
Step 4.4 review is mandatory before push, even when gate passes cleanly

**Updated**:  update skill

Added new Step 4 (quality validation) with the same gate + batch fix flow
Old steps 4–7 renumbered to 5–8

**Updated**: [copilot-instructions.md]

Validate subagent routing now references step 4 of the update flow (was step 4.3)

### CODE REVIEW

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+> |---|------|-----------|-------|
+> | 1 | `{filename}.mcs.yml` | {dimension} | {issue description} |
+> | 2 | `{filename}.mcs.yml` | {dimension} | {issue description} |
+>
+> **Recommendation:** {recommendation}
+>
+> What would you like to do?
+> - **A** — Fix all flagged cases (I'll propose a fix for each one)
+> - **B** — Pick which ones to fix
+> - **C** — Push as-is and fix later
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: Good catch — the validate subagent shouldn't be prompting the user for A/B/C, that's the parent's job, fixed this

File: solutions/ess-maker-skills/.github/copilot-instructions.md [hunk 2]
```diff
@@ -313,6 +314,15 @@ JSON format", "something went wrong with Workday".
 "flightcheck", "flight check", "is my agent ready", "validate my environment",
 "check my setup", "pre-flight", "deployment readiness", "run validation".
 
+**Quality validation invocation:** When quality validation is requested on
+eval files — at step 4.3 of the eval create flow OR step 4 of the eval
+update flow — invoke `runSubagent` pointing the subagent to read
+`src/skills/evaluations/validate/SKILL.md` as its first action, passing the
+paths of the newly written eval files and the agent folder path. Wait for
+the subagent to return with its quality report and any fix flow results
+before continuing. Do NOT proceed to review and push until the subagent
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/.github/copilot-instructions.md:324]: Done. [copilot-instructions.md](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) now correctly says the parent applies fixes and re-invokes the subagent — not that the subagent returns "fix flow results".

File: solutions/ess-maker-skills/src/skills/evaluations/update/SKILL.md [hunk 3]
```diff
@@ -80,7 +80,47 @@ extensionData:
...
+## Step 4: Quality validation
+
+Run quality validation on the changed files by invoking the validate subagent,
+passing the paths of all modified or newly created `.mcs.yml` files and the
+agent folder path.
+
+**After the subagent returns, display its full quality report output directly
+to the user — the complete dimension table, per-category scores, and any ⚠️
+callout blocks listing flagged files. Do NOT summarize or compress the report.
+Paste the subagent's Step 3 output verbatim into the chat so the user sees it.**
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/update/SKILL.md]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 4]
```diff
@@ -547,20 +582,65 @@ Run `python scripts/checkpoint.py "before evaluation test set creation"` to save
...
 
-### 4.3 — Review before push
+### 4.3 — Quality validation
+
+Run quality validation on the generated files.
+
+**After the subagent returns, display its full quality report output directly
+to the user — the complete dimension table, per-category scores, and any ⚠️
+callout blocks listing flagged files. Do NOT summarize or compress the report.
+Paste the subagent's Step 3 output verbatim into the chat so the user sees it.**
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+---
+
+## Step 5 — Fix flow (if user chooses A or B)
+
+For each flagged case the user wants to fix, one at a time:
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: already fixed

File: solutions/ess-maker-skills/.github/copilot-instructions.md [hunk 6]
```diff
@@ -313,6 +314,16 @@ JSON format", "something went wrong with Workday".
 "flightcheck", "flight check", "is my agent ready", "validate my environment",
 "check my setup", "pre-flight", "deployment readiness", "run validation".
 
+**Quality validation invocation:** When quality validation is requested on
+eval files — at step 4.3 of the eval create flow OR step 4 of the eval
+update flow — invoke `runSubagent` pointing the subagent to read
```
Comment (nehaoss) [inline solutions/ess-maker-skills/.github/copilot-instructions.md]: `runSubagent` doesn't appear anywhere else in the codebase. Is this the actual tool/API name the Copilot runtime provides? If the real mechanism is different, this instruction will silently fail. Should reference the actual invocation method or link to docs.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 7]
```diff
@@ -583,20 +670,20 @@ Show the user a summary of what was generated and ask for confirmation:
...
 
-### 4.5 — Push
+### 4.6 — Push
 
 Run `python scripts/push.py` to push the evaluation test sets to Copilot Studio.
 The push script handles two-pass ordering automatically: parent EvaluationSet records
 are created first, then child EvaluationData records are linked via `parentbotcomponentid`.
 
-### 4.6 — Show summary
+### 4.7 — Show summary
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:653]: Steps were renumbered (4.3->4.7 shift) but line 667 still says `remind them again in the final summary (Step 4.6)`. That reference is now stale and should be **Step 4.7**.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 8]
```diff
@@ -547,20 +582,72 @@ Run `python scripts/checkpoint.py "before evaluation test set creation"` to save
...
+
+When the user responds:
+
+- **A (fix all)**: Apply proposed fixes to all flagged cases.
+- **B (pick some)**: Ask the user which case numbers to fix (e.g. `1, 3, 5`).
+  Apply fixes to only those cases.
+- **C (push as-is)**: Proceed to step 4.4.
+
+**Fix flow (A or B)**:
+1. Apply the proposed fix to each selected case (edit the file).
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: This says 'Apply the **proposed fix**' but the validator (`validate/SKILL.md`) never produces proposed fixes. Its output table only has `File | Dimension | Issue` columns with no fix column. Either the validator needs to output a 'Proposed Fix' per flagged case, or this should say the agent devises the fix based on the issue description.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 9]
```diff
@@ -547,20 +582,72 @@ Run `python scripts/checkpoint.py "before evaluation test set creation"` to save
...
+2. After all edits are done, show a summary of what changed:
 
-Show the user a summary of what was generated and ask for confirmation:
+   > **Fixed {n} case(s):**
+   >
+   > | # | File | Dimension | Before | After |
+   > |---|------|-----------|--------|-------|
+   > | 1 | `{filename}.mcs.yml` | {dimension} | `{old input}` | `{new input}` |
+
+3. Re-invoke the validate subagent on the updated files and display the new
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: 'Re-invoke the validate subagent on the updated files' - does 'updated files' mean only the files just edited, or all files in the category? Category-level dimensions (Coverage, Redundancy, Diversity) need the **full set** to score correctly. If only edited files are passed, those scores will be meaningless. Should be explicit about scope.

File: solutions/ess-maker-skills/src/skills/evaluations/update/SKILL.md [hunk 10]
```diff
@@ -80,7 +80,54 @@ extensionData:
 For new test cases, create new `.mcs.yml` files following the naming convention:
 `{set-name}-{short-slug}.mcs.yml`.
 
-## Step 4: Review before push
+## Step 4: Quality validation
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/update/SKILL.md:83]: This entire quality validation + A/B/C fix flow block (~47 lines) is a near-verbatim copy of `create/SKILL.md` step 4.3. The repo's own 'No duplicate functions' rule in copilot-instructions.md applies here. If the flow logic changes, both files need syncing. Consider extracting this into `validate/SKILL.md` itself or a shared include so there's one source of truth.

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+folder. Use your file-reading tools to load each file cold and score it.
+
+---
+
+## Step 1 — Read the files
+
+Read each YAML file at the provided paths. For EvaluationData files, extract:
+- `input` — the user utterance being tested
+- `expectedOutput` — the expected agent behavior
+- `kind` — EvaluationData or MultiTurnEvaluationCase
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: `MultiTurnEvaluationCase` is listed here as something to extract, but all 9 scoring dimensions below are written for single-turn `input`/`expectedOutput` pairs. Multi-turn cases have a conversation-turn structure. The validator needs scoring guidance for multi-turn cases, or should explicitly state it skips them.

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+> |-----------|-------|----------------|
+> | Validity | {n}/5 | Inputs are grammatically correct and plausible as real user utterances |
+> | Realism | {n}/5 | Inputs sound like things real employees would say, not formal policy language |
+> | Assertion Quality | {n}/5 | Expected outputs are specific and describe observable agent behavior |
+> | Coverage | {n}/5 | Cases span a meaningful spread of sub-topics and positive/boundary/negative types |
+> | Diversity | {n}/5 | Inputs use genuinely different vocabulary, structure, and formality levels |
+> | Redundancy | {n}/5 | No two cases test the exact same input and expected behavior |
+> | Failure Mode Coverage | {n}/5 | Negative/edge cases reflect realistic failure modes, not contrived refusals |
+> | Discriminative Power | {n}/5 | Inputs are clearly scoped so they won't accidentally trigger the wrong topic |
+> | Topic Alignment | {n}/5 *(if applicable)* | Each case matches what its corresponding topic actually does |
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: Topic Alignment is marked `*(if applicable)*` here, and line 27 says it applies only to `topic-triggering`, `integration-data`, `multi-turn`. But there's no instruction on what to do for other categories - omit the row? Show 'N/A'? This ambiguity will produce inconsistent reports across runs.

File: solutions/ess-maker-skills/.github/copilot-instructions.md (hunk 6 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/.github/copilot-instructions.md]: runSubagent IS the real VS Code Copilot Chat tool — it's defined in the system-level agent framework (the <agents> block injected into the agent's system prompt by VS Code). It's not invented. We validated it works — the subagent appeared as "Explore: {description}" in the VS Code UI during testing.

Refer: https://code.visualstudio.com/docs/agents/subagents

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 7 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:653]: done, fixed

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 8 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 9 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/update/SKILL.md (hunk 10 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/update/SKILL.md:83]: fixed.

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md (hunk 11 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: yes removing multi turn for now from quality validation

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 13]
```diff
@@ -267,6 +267,29 @@ and negative variants (≥1 of each type per topic):
...
+| **Keyword** | Short, sparse input with no grammar | "open tkts" |
+
+Rules:
+- **Never generate two natural-language paraphrases for the same intent.** Near-synonyms
+  like "Show me my email" / "What's my email address on file?" test the same routing and
+  inflate redundancy scores without adding coverage.
+- **Positives**: if a topic gets 2 positive cases, make one natural language and one
+  keyword (e.g., "List my open IT tickets" + "open tkts").
+- **Negatives**: vary BOTH the failure-mode type AND the utterance type across topics.
+  Do not write all negatives as the same "Show me [person]'s [X]" natural-language
```
Comment (GrahamMcMynn) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: Why is this in here?  you have a negative section, isnt' this redundnat?

**Review by GrahamMcMynn — DISMISSED** [review]

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md (hunk 12 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: fixed, Done. *(if applicable)* removed from the table row, and an explicit instruction added: omit the row entirely for categories other than topic-triggering, [integration-data](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html), multi-turn. 

And, Done. multi-turn removed from all three places in the validate SKILL. Topic Alignment now applies to topic-triggering and [integration-data](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) only.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 13 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: good catch, fixed it, thank you!!

**Review by GrahamMcMynn — APPROVED** [review]
